VeloC plugin (#661)
pdidev/pdi: plugins/veloc/README.md

# The VeloC plugin {#veloc_plugin}

The VeloC plugin enables  

* writing and reading generic-data checkpoint files using [VeloC memory-based API](https://veloc.readthedocs.io/en/latest/api.html#example)

> * writing and reading applications' custom checkpoint files using [VeloC file-based API](https://veloc.readthedocs.io/en/latest/api.html#example#memory-based-api)

Author: The first one is correct while the second one should be  https://veloc.readthedocs.io/en/latest/api.html#memory-based-api. 

Is this what you meant?

The VeloC plugin does not currently support the full set of features of VeloC, but it offers a simple
declarative interface to access the core features, as explained in the following paragraphs. 

## External Dependencies   
In order to use the VeloC plugin, the user needs to install VeloC 1.8 or above as it is not vendored in PDI. This can be done following the instructions [here](https://veloc.readthedocs.io/en/latest/quick.html#download-and-install). <br>

The user also needs an MPI implementation, such as 
  - [openmpi](https://www.open-mpi.org/) 4.1 or above,
  - [mpich](https://www.mpich.org/) 4.3.2 or above with --with-ch4-shmmods configure options to enable shared memory support. Please note the [debian mpich package](https://tracker.debian.org/pkg/mpich) for versions above 4.3.2 does not build mpich with shared memory support and is therefore not suitable for the use of the plugin. 
## Configuration grammar

### Mandatory Mappings

The VeloC plugin specification tree requires 3 mandatory mappings for successful initialization. 

> * *config_file* : path to [VeloC's own configuration file](https://veloc.readthedocs.io/en/latest/userguide.html#setup). <br> Please note the plugin currently only works in synchronous mode. This means the parameter "mode" should be set to "sync" in VeloC's configuration file when using the plugin. 

Reviewer: ```suggestion
* *config_file* : path to [VeloC's own configuration file](https://veloc.readthedocs.io/en/latest/quick.html#configure). <br> Please note the plugin currently only works in synchronous mode. This means the parameter "mode" should be set to "sync" in VeloC's configuration file when using the plugin. 
```

> * *config_file* : path to [VeloC's own configuration file](https://veloc.readthedocs.io/en/latest/quick.html#configure). <br> Please note the plugin currently only works in synchronous mode. This means the parameter "mode" should be set to "sync" in VeloC's configuration file when using the plugin. 

Author: I do not. I can implement a  check

> * *config_file* : path to [VeloC's own configuration file](https://veloc.readthedocs.io/en/latest/quick.html#configure). <br> Please note the plugin currently only works in synchronous mode. This means the parameter "mode" should be set to "sync" in VeloC's configuration file when using the plugin. 

Author: It is not so straightforward to have a check on this since VeloC does not provide a getter-type function to query the mode selected by the user. I would have to parse the veloc configuration file myself but I could not use paraconf since this is not a yaml file. 

To be discussed. For now I think it is best to highlight in bold in the README that it should not be used in async mode.

```yml
config_file: ./veloc_config.cfg 
```

> * *checkpoint_label*: label used by VeloC to name checkpoint files. When using the VeloC plugin to checkpoint an application, the user may set the *checkpoint_label* equal to "myapp". If a failure occurs and the user wishes to restart the simulation from a checkpoint, the specification tree used in the new simulation run must have the *checkpoint_label* also equal to "myapp". 

Reviewer: ```suggestion
* *checkpoint_label*: label used by VeloC to name checkpoint files. When using the VeloC plugin to checkpoint an application with *checkpoint_label* equal to "myapp", the user must use the same label to restart the simulation from the checkpoint. 
```

```yml
checkpoint_label: myapp
```

> * *iteration*: name of the data corresponding to the simulation loop's iterator in the PDI data store. This is also used by VeloC to name checkpoint files. 

Author: `Iteration` is the keyword I chose for the plugin. I think in the VeloC documentation they refer to it as "version" but I found it a bit misleading.

> * *iteration*: name of the variable corresponding to the simulation loop's iterator in the PDI data store. This is also used by VeloC to name checkpoint files. 

Reviewer: ```suggestion
* *iteration*: identifier of the data corresponding to the simulation loop's iterator in the PDI data store. This is also used by VeloC to name checkpoint files. 
```

Author: Doesn't "identifier" make users think of an int? While this field should be a string (e.g. `iteration : iter_id` ) 

OK for "variable" --> "data"

```yml
iteration: ii
```

### Optional Mappings 

> * *status* : name of an integer variable in the PDI data store that represents whether a recovery is needed (status = 0) or not (status = 1). 

Reviewer: ```suggestion
* *status* : identifier of an integer data in the PDI data store that represents whether a recovery is needed (status = 0) or not (status = 1). 
```

Author: Same question as above

```yml
status: checkpoint_status
```
The plugin by default defines the status equal to 1 but users can change the status by writing to PDI.
```cpp 
PDI_expose("status", &checkpoint_status, PDI_OUT);
``` 
If users wish to inspect the status they can read it from PDI. 
```cpp 
PDI_expose("status", &checkpoint_status, PDI_IN);

> ``` 

Reviewer: Add an indent before so this content appears under the item `status`, rather than at the same level

* *counter* : name of an integer variable in the PDI data store that represents the number of checkpoints written by the plugin.

```yml
counter: checkpoint_counter
```

If users wish to inspect this value they can read it from PDI. 
```cpp 
PDI_expose("counter", &checkpoint_counter, PDI_IN);

> ```

Reviewer: add indent

### Checkpointing Behaviors

The user can define a *managed_checkpointing* tree or a *custom_checkpointing* tree. 

#### Managed Checkpointing
*managed_checkpointing* requires the following mappings: 

* *protected_data* : list of data structures to be checkpointed during a checkpoint event or to be recovered during a recover event (specify the same names as defined in the PDI data store). This list must include the simulation's iteration counter.

* *checkpoint_on_event* : name of the PDI event where the user wants to checkpoint

* *recover_on_event* : name of the PDI event where the user wants to recover

* *synchronize_on_event* : name of the PDI event where the user wants to synchronize. If the *status* is equal to 0, the synchronization event will be a recovery event, if the *status* is equal to 1, the synchronization event will be a checkpoint event.

> 

Author: You can have a synchronization event, a checkpoint event, and a recovery event in one simulation but they will all handle the same `protected_data`. 

Does that answer your question?

* *when* : PDI expression indicating when to execute a checkpoint operation. It is evaluated both in the case of a checkpoint event and of a synchronization event that performs a checkpoint. 

*managed_checkpointing* accepts the following optional mapping

>   * *recover_from_iteration* : The iteration from which the user wants to restore. For example, if the user sets this key to 20, the plugin will restore the checkpoint written at iteration 20.

Author: When this key is not defined, the checkpoint of the latest iteration is recovered. I can add this phrase.

```yml
managed_checkpointing:
        protect_data: [iter, main_field,elapsed_offset] # data to be checkpointed/recovered, iter must be included
        synchronize_on_event: newiter # name of PDI_event where to check the status and consequently checkpoint/recover  
        when : '$iter % 1000 = 0' 
        recover_from_iteration : 3000
```

#### Custom Checkpointing
*custom_checkpointing* requires the following mappings:

> * *veloc_file* : name of a char array variable in the PDI data store with a maximum size equal to 256. This variable is used by VeloC to map an input filename to the actual routed file path managed by VeloC.

Author: Yes it works with "./subdir/filename". 

OK -- for suggested change
